Simplify the definition and use of custom index operators

Octachron opened the pull request:
This patch simplifies the definition and use of custom index operator, i.e the
operators used to access or modify the value of an array-like structure at an given index.

First, a new syntax for index operators is introduced.
Six "core parenthesis operators" are added 
- `.()` ,
- `.[]` ,
- `.{}`, `.{,}`, `.{,,}`, `.{,..,}`.

The `.{,}`/`.{,,}`/`.{,,,}` operators are defined for compatibility with the
Bigarray syntax extension (more about it later). 
Each core index operator is available in an ~~safe/unsafe and~~ access/assignment
versions. For instance, `.()` is declined in
- `.()` : ~~safe~~ index operator
-   <del> .()! : unsafe index operator</del>
- `.()<-` : ~~safe~~ indexed assignment operator
- <del>.()!<- : unsafe indexed assignment operator</del>

Expressions of the form array.(index), string.(index) and bigarray1.{index} are
now desugared to one of these new index operators.  ~~The safe or unsafe index
operator are chosen in function of the presence or absence of the
"unsafe" compiler option.~~ These new index operators respect the `-unsafe`
compiler option using new compiler primitives.

In order to keep the compatibility with existing code,
the `.()` and `.[]` index operators are defined inside the pervasives modules
to be equal to their standard equivalent.
# Compatibility issues with bigarray

The bigarray `.{}` operators can be desugared to 4 different operators to
respect the original bigarray syntax extension.

More precisely, the index operator used in the desugaring changes if the 
index is a n-tuple. Depending on the arity of the tuple, the index operator
used is either: 
- 1-tuple => `.{}` : arity 2 [ +1 for assignment ] operator 
- 2-tuple => `.{,}` : arity 3 [ +1 for assignment ] operator
- 3-tuple => `.{,,}` : arity 4 [ +1 for assignment ] operator
- 4 and more tuples => `.{,..,}` : arity 2 [ +1 for assignment]
  operator, the tuple is passed as an array.

The bigarray module have been updated to use these new operators. However,
since these operators are not defined inside pervasives, this change **will 
break** existing code using bigarrays without opening the Bigarray module.
At least one test in the test suite has been updated due to this
incompatibility.

An other possibility to deal with this incompatibility issue
would be to hide the new desugar transform of `.{}` behind a compiler switch. 
A possible advantage of this compiler switch option is that it would 
allow to fusion all four `.{}` operator. Then, the bigarrays custom `.{}` index 
operators could be implemented at the Bigarray.Array{n} level rather than
the Bigarray level.   
# Other compatibility issues

Code relying on the old desugaring to redefine the index operators will also break.
# Example of code

``` ocaml
type z = Zero
type 'a succ=Succ
type (_,_) nat = Z : ('a,'a) nat | S : ('a,'b) nat -> ('a, 'b succ) nat

type  'dim vector = ((z,'dim) nat * float array )

let rec to_int: type inf sup. (inf,sup) nat -> int = 
function Z -> 0 | S n -> 1  + to_int n 

exception Wrong_dimension
let from_array : type dim. (z,dim succ) nat -> float array -> dim vector = 
fun (S nat) array ->  if (Array.length array=1+to_int nat) then 
 (nat, array) 
else
 raise Wrong_dimension

let ( .{} ) : 'dim vector -> ('mid,'dim) nat ->  'float  =
 fun (dim,v) nat -> Array.unsafe_get v (to_int nat)

let ( .{} <- ) : 'dim vector ->  ('mid,'dim) nat -> float -> unit  =
fun (dim,v ) nat x -> Array.unsafe_set v (to_int nat) x

let a = from_array (S (S Z)) [|1.;2.|]

let x = a.{Z}
let () =  a.{S Z} <- 11.
```

Edited on 13/06/2014 to reflect the current version of the proposal

Octachron committed:
Add a special syntax for index operators

This commit introduces a new syntax for index operators.
Six core parenthesis operator are added:
.(), .[], .{}, .{,}, .{,,}, .{,..,}.
The .{,}/.{,,}/.{,,,} operators are defined for compatibility with the Bigarray syntax extension.
Each core index operator is available in a access/assignement versions. For instance, .() is declined in
* .() : index operator
* .()<- : indexed assignment operator
The general syntax for these index operators as implemented in the parser is index_operator::= index_operator_core [<-]

Octachron committed:
Simplify the definition of custom .{} operators

This commits modify the Bigarray syntax extension in order to facilitate the use of custom .{} operators. The compatibility with the existing Bigarray syntax has been preserved as much as possible. However, this commit will break code which use the Bigarray .{}
syntax without opening the Bigarray module first!

Like the previous commit, this commit modifies the parser to desugar bigarray1.{index} to ( .{} ) bigarray1 index. Following the bigarray syntax, the index operator used
in the desugaring changes if the index is a n-tuple:
1-tuple => .{}
2-tuple => .{,}
3-tuple => .{,,}
4 and more tuples => .{,..,}

The bigarray modules has been modified to use this new index operators. Note that this means that these index operators are not anymore accessible without opening the bigarray module.

Octachron committed:
Simplify the use of custom .() and .[]

This commit modifies the parser to use the newly defined .() and .[] operators. It also moves the definition of the standard .() and .[] operator for String/Bytes and Array to the pervasives module.

Before this commit, expressions of the form array.(index) and string.(index) where desugared to Array.get[_unsafe] array index and Strinf.get[_unsafe] string index. The unsafe or unsafe version were chosen depending on the presence of the "-unsafe" compiler option. Such expression are now desugared to ( .() ) array index and ( .[] ) string index respectively. The same desugar operation is applied to array.(index) <- value which becomes ( .()<- ) array index value.

In order to keep the standard semantic for the string and array index operations, these new index operators are defined in the pervasives module using new compiler primitives, e.g.
let .() = "%array_opt_get".
These new primitives are then mapped to safe or unsafe version depending on the
the "-unsafe" compiler option. Consequently, these modifications should have no impact on existing code.

With these modifications, defining custom .() and .[] operators should be easier, at the cost of losing access to the standard index operator for either array or string.

hcarty commented on stdlib/pervasives.mli:
```diff
@@ -563,6 +563,45 @@ val ( @ ) : 'a list -> 'a list -> 'a list
+
+external  ( .() ) : 'a array -> int -> 'a = "%array_safe_get"
+(** Bound-checked parenthesis index operator.
+ a.(index) is desugared to ( .() ) a index if the compiler option -unsafe is not specififed *)
```
The desugared examples here and and in the following ocamldoc comments should probably be wrapped in `[]` so that it is properly marked up by ocamldoc:

```
... is desugared to [( .() ) a index] if the ....
```

thomasblanc commented on parsing/parser.mly:
```diff
@@ -149,8 +149,8 @@ let mkexp_constraint e (t1, t2) =
-let array_function str name =
-  ghloc (Ldot(Lident str, (if !Clflags.fast then "unsafe_" ^ name else name)))
+let array_function par assign=
+  ghloc ( Lident (par^assign))
```
Here the assign variable is a string, for bigarray_function it is a boolean.

Is there a reason for that ?
Maybe we should have a unified way of calling those two functions, the boolean way seems cleaner.

Octachron replied:
No good reason as far as I can see.
I was probably initially mimicking the original function, then realized that there was only two possible value for `assign` while writing `bigarray_function`.

thomasblanc commented on bytecomp/translcore.ml:
```diff
@@ -138,7 +138,16 @@ let comparisons_table = create_hashtable 11 [
+
+let caml_ba_indexop_ref n="%caml_ba_indexop_ref_"^(string_of_int n),   Pbigarrayref(!Clflags.fast, n, Pbigarray_unknown, Pbigarray_unknown_layout)
+let caml_ba_indexop_set n="%caml_ba_indexop_set_"^(string_of_int n), Pbigarrayset(!Clflags.fast, n, Pbigarray_unknown, Pbigarray_unknown_layout)
+
```
I tested the above lines and it wouldn't work: Translcore is linked before the arguments are evaluated, so in this case, !Clflags.fast is always set to false.

Those should go directly in the find_primitive function below.

Octachron replied:
You are right. I have fixed this problem in the current version. Thank you for taking the time to test this point.

alainfrisch commented:
The backward incompatible change (requiring to open Bigarray when the index operators are used) is indeed somewhat unusual: source-level compatibility for features that have been stable for many releases is usually expected.  I've opened a Mantis ticket ( http://caml.inria.fr/mantis/view.php?id=6765 ) with two proposals to help users in the transition: either make the error message more informative, or arrange so that a missing "open Bigarray" only results in a warning, not a fatal error (for a transition period, e.g. one major release).

Octachron committed:
Warnings for using .{} without opening Bigarray

The previous change to the index operator .{} (and friends) broke backward
source code compatibility with the previous version of Ocaml. Program using
these operators without opening the Bigarray module triggered an error.
To alleviate this problem, this patch adds a temporary hack to the type
checker to transform the corresponding errors in a warning.

dbuenzli commented:
The workaround for Bigarrays is very ugly, it brings a lot of identifiers in scope. If that's really the way people want to proceed please put that in a submodule to open.

ghost commented:
I just updated the printer for this change and I noticed that these expressions are all parsed the same way:

``` ocaml
x.{1,2}
x.{(1, 2)}
x.{begin 1, 2 end}
```

This was already the case before, but now that operators can be redefined and indexes can be of arbitrary types it is slightly surprising. It shouldn't be too hard to fix in the parser.

Octachron commented:
I didn't think about updating this part of the printer, sorry. Concerning the problem of tuple parsing, it's true that the behavior is surprising. However, updating it might break existing code at least theoretically; even if I hope that no code base relies on this particular feature of the parser. Anyway, I will try to propose a patch to fix this behavior to at least launch a discussion on this subject.

gasche commented:
I'm sorry to bring bad news, but we had a developer meeting today where we decided to revert this patch from trunk.

The justification is as follows: @lpw25 argued for an incompatible use of the `.(..)` syntax, where it is resolved by type-directed disambiguation (like record field projects today) instead of environment lookup. His proposal is not ready to be integrated in the next release, but it casts doubt on the wisdom of integrating this one in the next release (besides the other trouble we had with the bigarray breakage ( #248 and #299 ). He proposed to use his disambiguation technique for `.(..)` and leave `.[...]` scope-defined, but that just sounds fishy.

I would be happy this feature to be maintained in a branch of the compiler, but it will not be included in the next release to give us more freedom in the design space in the coming releases.

Octachron commented:
Thank you for letting me know. 

I must admit it is little disheartening to see this patch reverted due to a conflict with an undisclosed potential design that might be integrated in few years. The information asymmetry is quite irksome. But I understand that errors in design space are paid dearly and must be avoided at all cost.

> He proposed to use his disambiguation technique for .(..) and leave .[...] scope-defined, but that just sounds fishy.

That also sounds fishy to me. If the type disambiguation design is better for `.(..)`, why deprive `.[..]` from this improvement?

> (besides the other trouble we had with the bigarray breakage ( #248 and #299 )).

Troubles that could be solved by agreeing on one submodule name for the bigarray indexing operators and integrating a patch that have been sleeping on mantis for eight months without any reactions? 

> I would be happy this feature to be maintained in a branch of the compiler,

I am at a loss. Are you speaking of an officially supported branch or are you asking me to maintain this patch indefinitely with no visibility at all on any potential integration in trunk? Note that I might do that, but it would be more to experiment with better multidimensional index operators.

> but it will not be included in the next release to give us more freedom in the design space in the coming releases.

I wish the core team good luck in finding a nice extremum in the design space. I could really use a readable syntax for multidimensional array slicing à la numpy.

gasche commented:
Sorry for the terse writing above (I was trying to give feedback as soon as possible to the many change proposals that have been discussed, but evidently did a bad job at communicating here). 

There is nothing secret about Leo's proposal, except that he prepared some slides/presentation for the meeting that are not online yet ( @lpw25 , could you make them publicly available somewhere?). It is work that originated from an effort to remove the special-casing of float arrays in the current OCaml runtime. You can find some early description of the idea [in this comment](https://www.lexifi.com/blog/about-unboxed-float-arrays#comment-1255) on a proposal by Alain Frisch related to float arrays. Leo then developed a more complete proposal of what he calls "array data types", whose implementation can be found [in his array-data-types branch](https://github.com/lpw25/ocaml/commits/array-data-types), where a new type constructor for arrays is introduced with attributes to provoke various kind of specializations (mutable/immutable, unboxing, zero-termination, with the default being generic, non-float-specialized arrays).

The key point of his proposal that is of concern here is to consider these specialized array types as glorified records, where the array-indexing operation `.()` and `.()<-`, and a length projection `.length` have a status similar to record field projections -- in particular, they are disambiguated in a type-directed way, instead of being overloaded by scope as in your proposal.

It is not yet clear that this aspect of the proposal is bound to stay, but I think it is interesting enough (obviously for record projections type-based disambiguation is much more usable than scope-based usage, and as the complex discussion on shadowing warnigns show, local `open` are no panacea) to give us pause and some time to think harder about where we want to take indexing operators.

Having `.()` type-disambiguated and `.[]` scope-disambiguated is a proposal of Leo, and honestly I think it makes no sense. If we find out that the type-directed strategy is indeed better, it will surely be better for `.[]` as well (think of `bytes` vs `string` for example), and conversely if scope-based is a better design in terms of user applicability, then I think this aspect of Leo's proposal should be rejected. We are trying to find a good design, not play politics and make compromises to please the involved parties. (I'm not saying I would be against _all_ hybrid designs, but there would need to be a very strong justification for mixing both strategies.)

We are trying to move to a shorter release cycles, ideally six-month releases. The choice here is not to definitely reject your feature (scope-based user-defined indexing overloading), but rather to _not_ include it in the next release (4.03) and postpone an inclusion choice to 4.04 -- hoping to have advanced in the consideration of these competing designs. Language design is a conservative affair.

When I speak of a branch, I am thinking of having a `scoped-indexing-operators` branch in the development repository, that would be synchronized with `trunk` -- so that at any point in time it is easy to build a version of the distribution with the feature enabled, for experimentation or for specialized usage. I'm not asking you to maintain it -- but I'm not promising either that I will personally have the time to do so.

(Let me point out that the Bigarray-related compatibilty woes were not instrumental in taking this decision. Nor are their your responsibility, it's clear that at the time the feature was integrated there was consensus that not special-casing Bigarray was the right choice -- but this was proved wrong by practical experience with trunk compatibility. In fact we had discussed them separately before taking this reversion decision, and had decided to go the conservative way of implicitly opening `Bigarray.Operators` an `.{..}` operator was used with none in scope.)

lpw25 commented:
>  @lpw25 , could you make them publicly available somewhere?

I didn't have any actual slides, but the branch is available at https://github.com/lpw25/ocaml/tree/array-data-types. It doesn't have any documentation to explain the syntax, since I haven't added tests yet, so that may not be very useful for understanding the feature.

> Having .() type-disambiguated and .[] scope-disambiguated is a proposal of Leo, and honestly I think it makes no sense.

I disagree. Whilst the current patch doesn't do this yet, it is sufficient to use ".()" for both regular arrays and strings (and bytes etc.), in fact that is a benefit of the proposal: ".()" becomes an explicit projection from an array(-like) block. However, it means that ".()" is always restricted to types which are implemented as array(-like) blocks. Since many other types may usefully have some notion of projection it makes sense to also separately provide an operator that can be over-loaded for those types and it seems reasonable for ".[]" to be that operator.

lpw25 commented:
As a side-note, my array data types proposal -- in common with ordinary labels -- supports the syntax:

```
foo.FloatArray.(0)
```

which avoids the need for local open. I think a similar syntax would be useful for ".[]".

Drup commented:
I have trouble to understand how you can't do all that with the current proposal + implicits (and an Indexable signature), and without more insight, I don't see how @lpw25's proposal is likely to get integrated before implicits.

lpw25 commented:
> I have trouble to understand how you can't do all that with the current proposal + implicits (and an Indexable signature)

It's difficult to explain without fully explaining the proposal -- so best to wait until I've found time to do that.

> I don't see how @lpw25's proposal is likely to get integrated before implicits.

The proposal is already mergeable whilst implicits still requires a huge amount of work.

Octachron commented:
@gasche: 

Thanks for taking the time to detail your comment. I have no doubt that you were well meaning in your previous comment. However, communicating decisions without the context of the decision is bound to create some frictions from time to time. If there is any plan to have some publicly available minutes of the developer meeting? It would be very nice; the difference between a secret and an information shared only with a select circle is tenuous at best.

After reading the explanation for Leo proposal, it sounds like a nice feature; that does not cover any of my uses case for custom index operators (e.g dictionary, slice, multidimentionnal array or statistically checked array ) where either the data is not exactly an array block or the index is not a simple integer. In these more complex situations, I still think that it makes more sense to have scope-overloaded index operators. So I agree with @lpw45 that it might make sense to have `.()` for "true" array  type and `.[]` for indexable type ( and `.{}` for use cases that match bigarray semantic accidentally ). 

> We are trying to move to a shorter release cycles, ideally six-month releases. The choice here is not to definitely reject your feature (scope-based user-defined indexing overloading), but rather to not include it in the next release (4.03) and postpone an inclusion choice to 4.04 -- hoping to have advanced in the consideration of these competing designs. Language design is a conservative affair.

The shorter release cycles sounds like a good news. This is also the kind of news for which having minutes would be nice. And like I said before, I understand the thirst for further considerations -- just does not expect me to be perfectly happy if these considerations take place behind closed doors.

> When I speak of a branch, I am thinking of having a scoped-indexing-operators branch in the development repository, that would be synchronized with trunk -- so that at any point in time it is easy to build a version of the distribution with the feature enabled, for experimentation or for specialized usage. I'm not asking you to maintain it -- but I'm not promising either that I will personally have the time to do so.

In this context, i.e. limited period and well-defined objective, I will be glad to lend an hand to maintain this branch.

@lpw45:

I agree that the syntax `array.Module.{..}` is a nice way to avoid both local open and custom ascii art operators. Originally, I was thinking to propose this extended syntax after 4.03. Let's say that I am a little bit wary about such extensions now.

gasche commented:
> If there is any plan to have some publicly available minutes of the developer meeting?

I am currently in the process of "exporting" the decisions that have been made as publicly-visible comments on the tickets that have been discussed. I did a bit yesterday (obviously I over-favoured quantity over quality of the reports), and hope to finish today. (It's actually quite a bit of work.)

We do need to improve communication on release cycle decisions and such (I'm trying to mention it explicitly in the present reports), but I'm uncomfortable with the idea of exporting minutes for a meeting without having had agreement of the various participants beforehand. For this one meeting I think I'll keep a summary-only approach. (I don't expect to be present in similar meetings in the short- to medium-term future, but maybe @bobot or other participants could take a decision on this. )

lpw25 commented:
> but I'm uncomfortable with the idea of exporting minutes for a meeting without having had agreement of the various participants beforehand.

I think a summary of the discussions (rather than full minutes) would be useful in addition to the individual feedback to PRs as it gives a single place people can read about the various discussions. I also agree that the agreement of the participants is needed before releasing such a summary. Perhaps if you wrote a small summary from your notes, and asked on caml-devel for the approval of the participants, they could be posted somewhere (although I'm not sure where the best place would be).

alainfrisch commented:
The minutes would also be useful for caml-devel itself since not all core developers could attend the meeting.  For a larger diffusion, we would also need to get the green light from non-core devel participants.

mshinwell commented:
@damiendoligez As far as I recall from the recent meeting it was decided that this should not be merged; do I remember correctly?  If so we should probably close this.

Octachron commented:
@mshinwell I have more mixed memories on this discussion, however it would be clearer to close this PR (and #622) in favor of its latest incarnation #1064 .

damiendoligez commented:
My notes agree with @Octachron
